ui-src/lib/views/CalendarView.js

```diff
@@ -0,0 +1,418 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// In addition, as a special exception, the copyright holders give
+// permission to link the code of portions of this program with the
+// OpenSSL library under certain conditions as described in each
+// individual source file, and distribute linked combinations including
+// the two.
+//
+// You must obey the GNU General Public License in all respects for all
+// of the code used other than OpenSSL. If you modify file(s) with this
+// exception, you may extend this exception to your version of the
+// file(s), but you are not obligated to do so. If you do not wish to do
+// so, delete this exception statement from your version. If you delete
+// this exception statement from all source files in the program, then
+// also delete it here.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import $ from 'jquery';
+import 'jquery-ui/themes/base/button.css';
+import 'jquery-ui/themes/base/core.css';
+import 'jquery-ui/themes/base/datepicker.css';
+import 'jquery-ui/themes/base/dialog.css';
+import 'jquery-ui/themes/base/resizable.css';
+import 'jquery-ui/themes/base/theme.css';
+import 'jquery-ui/themes/base/tooltip.css';
+import 'jquery-ui/ui/widgets/datepicker';
+import 'jquery-ui/ui/widgets/dialog';
+import 'jquery-ui/ui/widgets/tooltip';
+
+import DatePickerView from './DatePickerView';
+import CalendarTSRange from '../models/CalendarTSRange';
+import {TimeStamp90kFormatter} from '../support/TimeFormatter';
+import Time90kParser from '../support/Time90kParser';
+
+/**
+ * Find the earliest and latest dates from an array of CameraView
+ * objects.
+ *
+ * Each camera view has a "days" property, whose keys identify days with
+ * recordings. All such dates are collected and then scanned to find earliest
+ * and latest dates.
+ *
+ * "days" for camera views that are not enabled are ignored.
+ *
+ * @param  {[Iterable]} cameraViews Camera views to look into
+ * @return {[Set, String, String]}       Array with set of all dates, and
+ *                                       earliest and latest dates
+ */
+function minMaxDates(cameraViews) {
+  /*
+   * Produce a set with all dates, across all enabled cameras, that
+   * have at least one recording available (allDates).
+   */
+  const allDates = new Set(
+    [].concat(
+      ...cameraViews
+        .filter((v) => v.enabled)
+        .map((v) => Array.from(v.camera.days.keys()))
```

Comment:
In recent history this kind of functional style of programming has become the norm (where I live at least) and it is easily readable (easier than the original code), and expresses intent very well. The code itself executes only once per initial application load and is not on the critical path at all. So I am not worried to this kind of "optimization" and favor expressive code.

Reply:
OK, so with the exceptions commented on above, updated the PR. Also note that a future PR that I have ready will change eslint settings so that the misspelling example in DateViewPicker will be picked up. That should alleviate your concerns somewhat for the future.

Reply:
This code should execute once per camera checkbox click, right? Ehh, I suppose it's fine, but it seems odd to me to generate the extra intermediary products. We'll need to touch this code again soon anyway (when merging your UI with the `new-schema` branch, which adds the stream dimension here). I'll likely send you a PR for that merge, and we can discuss more then.

Reply:
Gets called on camera checkbox click, and also once after application load. Lets keep for now and see how it needs to be changed for the new schema.